### DISCUSSION TITLE
Plotting PB2NC output field

### DISCUSSION THREAD

**Discussion by BinbinZhou-NOAA** [dtcenter/METplus#1383] [Plot Generation] [answered]
I am wondering  why I can not find TMP/Z2 from my output file of  PB2NC. 
The GFS prepbufr was first converted to NC file by
PB2NC.   By using ncdump, I can see TMP in the output NC file.
But using plot_data_plane, the error says TMP is not found.
The sample case can be found on MARS in this directory
/gpfs/dell2/emc/modeling/noscrub/Binbin.Zhou/met/pb2nc

The script run_pb2nc.sh is run PB2NC
plot.sh is plotting

**Comment by JohnHalleyGotway** [comment]
@BinbinZhou-NOAA thanks for moving this issue over to discussions.

My guess is that the surface temperature (i.e. 2m TMP) obs are missing because of their quality flag. We have a note about this buried on the [Input Data page](https://dtcenter.org/community-code/model-evaluation-tools-met/input-data) of the MET website, at the very bottom.

**NOTE: The quality flag values for surface observations in GDAS PrepBufr files are set to a value of 9 to prevent them from being used in the data assimilation process. To retain them in the PB2NC tool, the quality mark threshold in the configuration file must be set to at least 9.**  

Perhaps you're using the [default quality_mark_thresh](https://github.com/dtcenter/MET/blob/5e1f551d86fcb5b731e3a83c0b043dbd7ef3c7aa/met/data/config/PB2NCConfig_default#L129) value of 2 and the actually quality mark of those surface obs are set to 9?

To check, please try rerunning METplus using:
```
PB2NC_QUALITY_MARK_THRESH = 9
```

Does that result in 2-m temperature obs in the output?

If not, I can go take a closer look at your data.

**Reply by BinbinZhou-NOAA** [reply]
I already set the quality_mark_thresh = 9; in the config file
/gpfs/dell2/emc/modeling/noscrub/Binbin.Zhou/met/pb2nc/PB2NCConfig
But still got error. This time I ttried to plot 4 fields (TMP, DPT, UGRD and VGRD), the errors are
ERROR  : plot_data_plane -> trouble getting field "name="TMP"; level="Z2";" from file "gfs.t00z.prebufr.f00.nc"
ERROR  :
DEBUG 1: Opening data file: gfs.t00z.prebufr.f00.nc
ERROR  :
ERROR  : plot_data_plane -> trouble getting field "name="DPT"; level="Z2";" from file "gfs.t00z.prebufr.f00.nc"
ERROR  :
DEBUG 1: Opening data file: gfs.t00z.prebufr.f00.nc
ERROR  :
ERROR  : plot_data_plane -> trouble getting field "name="UGRD"; level="Z10";" from file "gfs.t00z.prebufr.f00.nc"
ERROR  :
DEBUG 1: Opening data file: gfs.t00z.prebufr.f00.nc
ERROR  :
ERROR  : plot_data_plane -> trouble getting field "name="UGRD"; level="Z10";" from file "gfs.t00z.prebufr.f00.nc"

**Comment by malloryprow** [comment]
@BinbinZhou-NOAA I'm not sure surface fields are in the GFS prepbufr file. For my grid-to-obs verification, I use the NAM prepbufr files.

**Reply by BinbinZhou-NOAA** [reply]
In our EVS GEFS verification plan, the surface fields are verified against prepbufr over both global and CONUS

**Reply by malloryprow** [reply]
Yes. I think the decisions are still being made which exactly what prepbufr files are to be used for the grid-to-obs verification.

**Reply by BinbinZhou-NOAA** [reply]
I just tested nam's prepbufr, but got same error

**Comment by BinbinZhou-NOAA** [comment, marked as answer]
I have created a completed sample package on Mars that contains everything including 
input GFS prepbufr, system.conf and 2 pb2nc conf files. The directory is
/gpfs/dell2/emc/modeling/noscrub/Binbin.Zhou/met/PB2NC_test
There 2 tests to indicate 2 issues as below. Both tests are executed by the same script run_pb2nc.sh
Test1 uses gefs_pb2nc.conf,  
  in which all time (da_init) is hardcoded. the running is completed and output nc file is generated. 
  But if use plot_data_plan to draw plot for any fields, the result is "field not found"
Test2 uses gefs_pb2nc.da_init.conf
  in which time is set by {da_init?fmt=%Y%m%d}. The execution is failed. Error message in the log
file says 
    KeyError: 'da_init?fmt=%Y%m%d' 
The 2 tests use same prepbufr data in sub-directory input_data/ 
The output_base for these 2 tests are output_base_test1  and output_base_test2, respectively
If copy this package to other machine, the system.conf and the running script file should be modified for METplus_10.0.0 installed path.

**Reply by JohnHalleyGotway** [reply]
@BinbinZhou-NOAA sorry for not realizing this sooner. You're running the wrong tool. The plot_data_plane tool plots 2-dimensional fields of gridded data. The plot_point_obs tool plots the locations of point observations... i.e. the NetCDF point observation output files created by pb2nc, ascii2nc, madis2nc, and so on). You should be running plot_point_obs, not plot_data_plane.

I'd recommend thinking about what/how you'd like to plot the point observations. Below are 2 examples of output generated by the plot_point_obs tool, one simple and one complex.

<img width="1106" alt="Screen Shot 2022-01-28 at 9 58 19 AM" src="https://user-images.githubusercontent.com/21087144/151589197-54563720-9f66-4286-b20f-053a2f7f99a5.png">

Using the default settings, the image on the left just plots a small red dot for each observation (lat, lon) location it finds. In this case, it's plotting ADPUPA temperature observations using the "[-obs_var TMP -msg_typ ADPUPA](https://github.com/dtcenter/MET/blob/main_v10.0/test/xml/unit_plot_point_obs.xml#L40)" command line options, and by default, it does so on a global grid.

The image on the right is doing a lot more in this [test](https://github.com/dtcenter/MET/blob/5e1f551d86fcb5b731e3a83c0b043dbd7ef3c7aa/test/xml/unit_plot_point_obs.xml#L49), as specified by this [config file](https://github.com/dtcenter/MET/blob/main_v10.0/test/config/PlotPointObsConfig). It defines a 2D data field (500mb RH) as the base image over which 3 different types of observations are plotted. TRMM (ADPSFC APCP) observations are plotted as small blue dots. ADPSFC RH observations are plotted as small gray dots. And ADPUPA RH observations at 500mb are plotted as large circles whose color is plotted using the same color table as the base image.

So give the options for plot_point_obs some thought. If you want to display 4 observation variables, you could configure it to do so in one image with a config file. It just all depends on your goals.

**Comment by BinbinZhou-NOAA** [comment]
John,

  Thanks for pointing that out the correct tool plot_point_obs for point
observation data.
Yes, I can generate TMP/Z2 plot with the plot_point_obs config file.

[image: gfs.t00z.prebufr.f00.nc.png]

So the first issue the https://github.com/dtcenter/METplus/discussions/1383
is resolved. The second issue is that I asked you George days ago:
When I run pb2nc with  PB2NCConfig_wrapped, the error says
{da_init?fmt=%Y%m%d}  error
Last time, you copied the testing files for George, now I suggest you
copied a new sample
package that includes everything for this testing case. The package's  path
on Mars is
/gpfs/dell2/emc/modeling/noscrub/Binbin.Zhou/met/PB2NC_test
This sample is very simple, and  the running script is run_pb2nc.sh
See discussion#1383 for detail.

Thanks!

Binbin


On Fri, Jan 28, 2022 at 12:08 PM johnhg ***@***.***> wrote:

> @BinbinZhou-NOAA <https://github.com/BinbinZhou-NOAA> sorry for not
> realizing this sooner. You're running the wrong tool. The plot_data_plane
> tool plot 2-dimensional fields of gridded data. The plot_point_obs tool
> plots the locations of point observations... i.e. the NetCDF point
> observation output files created by pb2nc, ascii2nc, madis2nc, and so on).
>
> I'd recommend giving some thought to what/how you'd like to plot the point
> observations. Below are 2 examples of output generated by the
> plot_point_obs tool.
>
> [image: Screen Shot 2022-01-28 at 9 58 19 AM]
> <https://user-images.githubusercontent.com/21087144/151589197-54563720-9f66-4286-b20f-053a2f7f99a5.png>
>
> Using the default settings, it just plots a small red dot for each
> observation (lat, lon) location it finds. In this case, it's plotting
> ADPUPA temperature observations, and by default, it does so on a global
> grid.
>
> The image on the right is doing a lot more, as specified by this config
> file
> <https://github.com/dtcenter/MET/blob/main_v10.0/test/config/PlotPointObsConfig>.
> It defines a 2D data field (500mb RH) as the base image over which 3
> different types of observations are plotted. TRMM (ADPSFC APCP)
> observations are plotted as small blue dots. ADPSFC RH observations are
> plotted as small gray dots. And ADPUPA RH observations at 500mb are plotted
> as large circles whose color is plotted using the same color table as the
> base image.
>
> So give the options for plot_point_obs some thought. If you want to
> display 4 observation variables, you could configure it to do so in one
> image with a config file. It just all depends on your goals.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/METplus/discussions/1383#discussioncomment-2068047>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQBMPFLPKTOKB4CQ7UKUI5TUYLEPDANCNFSM5M7IHSUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Binbin Zhou

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct.

College Park, MD 20740

***@***.***

301-683-3683

**Comment by JohnHalleyGotway** [comment]
@BinbinZhou-NOAA, since this discussion is about plotting pb2nc output, I went ahead and selected an answer and will "Lock" it to prevent future posts.

Regarding your question about configuring PB2NC, please check your email for a response from @georgemccabe last week. You were using the filename templates in PB2NC_INPUT_DIR which doesn't work. If that doesn't resolve your issue, please post a new discussion, and I'll go retrieve the data for further investigation.

I realize this may be a hassle, but we're trying to keep each discussion as focussed as possible. That should make the history of Discussion more useful and clear to other users in the future.
